File: notebooks/bennink.py
# -*- coding: utf-8 -*-
"""
Created on Wed Feb 24 14:55:23 2016

@author: schiavon
"""

from util import lambda2omega
import numpy as np
from myintegrate import complex_integral
from sellmeier import n
from scipy.constants import c

def func(z,lp,ls,li,wp,ws,wi,axp,axs,axi,Lambda):
    omega_p = lambda2omega(lp)
    omega_s = lambda2omega(ls)
    omega_i = lambda2omega(li)
    
    def qp(z):
        return wp**2 + 1j*z*lp/np.pi/n(lp,axp,typ='christ')

    def qs(z):
        return ws**2 + 1j*z*ls/np.pi/n(ls,axs,typ='christ')
        
    def qi(z):
        return wi**2 + 1j*z*li/np.pi/n(li,axi,typ='christ')
    
    return wp*ws*wi*np.exp(1j*( (n(lp,axp,typ='christ')*omega_p - n(ls,axs,typ='christ')*omega_s - n(li,axi,typ='christ')*omega_i)/c + 2*np.pi/Lambda )*z) / (np.conj(qs(z))*np.conj(qi(z)) + qp(z)*np.conj(qi(z)) + qp(z)*np.conj(qs(z)))
#    return wp*ws*wi / (np.conj(qs(z))*np.conj(qi(z)) + qp(z)*np.conj(qi(z)) + qp(z)*np.conj(qs(z)))

def O(lp,ls,li,wp,ws,wi,axp,axs,axi,L,Lambda):
    return complex_integral(func,-L/2,L/2,(lp,ls,li,wp,ws,wi,axp,axs,axi,Lambda))

File: matlab/SPDC/bennink.m
function out = bennink( lp,ls,li,wp,ws,wi,axp,axs,axi,L,Lambda )
%BENNINK Summary of this function goes here
%   Detailed explanation goes here

    function out = func(z,lp,ls,li,wp,ws,wi,axp,axs,axi,Lambda)


        function out = qp(z)
            out = wp.^2 + 1i.*z.*lp./pi./n(lp,axp);
        end

        function out = qs(z)
            out = ws.^2 + 1i.*z.*ls./pi./n(ls,axs);
        end

        function out = qi(z)
            out = wi.^2 + 1i.*z.*li./pi./n(li,axi);
        end

        out = wp.*ws.*wi.*exp(1i.* 2.*pi.*(n(lp,axp)./lp - n(ls,axs)./ls - n(li,axi)./li + 1./Lambda ).*z) ./ (conj(qs(z)).*conj(qi(z)) + qp(z).*conj(qi(z)) + qp(z).*conj(qs(z)));
    end

    f = @(z) func(z,lp,ls,li,wp,ws,wi,axp,axs,axi,Lambda);
    out = integral(f,-L/2,L/2);


end

